server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,164 @@
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      val _ = ctx.writeAndFlush(res)
+    }
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        ctx.pipeline().remove(this)
+        ctx.pipeline().remove("readTimeoutHandler")
+        ReferenceCountUtil.release(msg)
+        val (runningFuture, _) = unsafeRunAsync { () =>
+          route(NettyServerRequest(req.retain()))
+            .map {
+              case Some(response) => response
+              case None           => ServerResponse.notFound
+            }
+        }
+
+        val _ = runningFuture.transform {
+          case Success(serverResponse) =>
+            try {
+              serverResponse.body match {
+                case Some(function) => {
+                  val content = function(ctx)
+                  content match {
+                    case r: ReactiveWebSocketProcessorNettyResponseContent => {
+                      ctx
+                        .pipeline()
+                        .addAfter(
+                          "serverCodecHandler",
+                          "wsControlFrameHandler",
+                          new NettyControlFrameHandler(
+                            ignorePong = r.ignorePong,
+                            autoPongOnPing = r.autoPongOnPing,
+                            decodeCloseRequests = r.decodeCloseRequests
+                          )
+                        )
+                      r.autoPing.foreach { case (interval, pingMsg) =>
+                        ctx
+                          .pipeline()
+                          .addAfter("wsControlFrameHandler", "wsAutoPingHandler", new WebSocketAutoPingHandler(interval, pingMsg))
+                      }
+                      // Manually completing the promise, for some reason it won't be completed in writeAndFlush. We need its completion for NettyBodyListener to call back properly
+                      r.channelPromise.setSuccess()
+                      val _ = ctx.writeAndFlush(
+                        // Push a special message down the pipeline, it will be handled by HttpStreamsServerHandler
+                        // and from now on that handler will take control of the flow (our NettyServerHandler will not receive messages)
+                        new DefaultWebSocketHttpResponse(
+                          req.protocolVersion(),
+                          HttpResponseStatus.valueOf(200),
+                          r.processor, // the Processor (Pipe) created by Tapir interpreter will be used by HttpStreamsServerHandler
+                          new WebSocketServerHandshakerFactory(wsUrl(req), null, false)
+                        )
+                      )
+                    }
+                    case otherContent =>
+                      logger.error(s"Unexpected response content: $otherContent")
```

Comment:
maybe extract these into methods as well, with names specifying what kind of error / misues occured, e.g. `serverErrorUnexpectedResponseContent` or `clientErrorNotAWSEndpoint` or sth like this. The main flow would read much nicer then